### ISSUE TITLE
A Concurrent-safe Centralized Pointer Managing Facility

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#9280]
A Concurrent-safe Centralized Pointer Managing Facility<br>
<br>
https://ift.tt/HrVPkbG<br>
<br>
<br>
<br>

<div><div><div><div>
<div>
<p>Author(s): <a href="https://golang.design/cdn-cgi/l/email-protection#1e6c7b6d7b7f6c7d763b2b5c7f6a3b2b5a7d767f7079756b70307a7b">Changkun Ou</a></p>
<p>Permalink: <a href="https://golang.design/research/cgo-handle"></a><a href="https://golang.design/research/cgo-handle">https://golang.design/research/cgo-handle</a></p>
<p>In the Go 1.17 release, we contributed a new cgo facility <a href="https://tip.golang.org/pkg/runtime/cgo/#Handle">runtime/cgo.Handle</a> in order to help future cgo applications better and easier to build concurrent-safe applications while passing pointers between Go and C. This article will guide us through the feature by asking what the feature offers to us, why we need such a facility, and how exactly we contributed to the implementation eventually.</p>
<h2>Starting from Cgo and X Window Clipboard</h2>
<p>Cgo<sup></sup> is the de facto approach to interact with the C facility in Go. Nevertheless, how often do we need to interact with C in Go? The answer to the question depends on how much we work on the system level or how often do we have to utilize a legacy C library, such as for image processing. Whenever a Go application needs to use a legacy from C, it needs to import a sort of C dedicated package as follows, then on the Go side, one can simply call the <code>myprint</code> function through the imported <code>C</code> symbol:</p>
<div><pre><code><span> 1</span><span>/*
</span><span> 2</span><span>#include &lt;stdio.h&gt;
</span><span> 3</span><span>
</span><span> 4</span><span>void myprint() {
</span><span> 5</span><span>	printf("Hello %s", "World");
</span><span> 6</span><span>}
</span><span> 7</span><span>*/</span>
<span> 8</span><span>import</span> <span>"C"</span>
<span> 9</span>
<span>10</span><span>func</span> <span>main</span>() {
<span>11</span>	C.<span>myprint</span>()
<span>12</span>	<span>// Output:
</span><span>13</span><span></span>	<span>// Hello World
</span><span>14</span><span></span>}
</code></pre></div>
<p>A few months ago, while we were working on building a new package <a href="https://golang.design/x/clipboard"><code>golang.design/x/clipboard</code></a><sup></sup>, a package that offers cross-platform clipboard access. We found out, there is a lacking of facility in Go, despite the variety of approaches in the wild, still suffering from soundness and performance issues.</p>
<p>In the <a href="https://golang.design/x/clipboard"><code>golang.design/x/clipboard</code></a> package, we had to cooperate with cgo to access system level APIs (technically, it is an API from a legacy and widely used C system), but lacking the facility of knowing the execution progress on the C side. For instance, on the Go side, we have to call the C code in a goroutine, then do something else in parallel:</p>
<div><pre><code><span>1</span><span>go</span> <span>func</span>() {
<span>2</span>	C.<span>doWork</span>() <span>// Cgo: call a C function, and do stuff on C side
</span><span>3</span><span></span>}()
<span>4</span>
<span>5</span><span>// .. do stuff on Go side ..
</span></code></pre></div>
<p>However, under certain circumstances, we need a sort of mechanism to understand the execution progress from the C side, which brings the need of
communication and synchronization between the Go and C. For instance, if we need our Go code to wait until the C side code finishes some initialization work until some execution point to proceed, we will need this type of communication precisely to understand the progress of a C function.</p>
<p>A real example that we encountered was the need to interact with the clipboard facility. In Linux's <a href="https://en.wikipedia.org/wiki/X_Window_System">X window environment</a>, clipboards are decentralized and can only be owned by each application. The ones who need access to clipboard information are required to create their clipboard instance. Say an application <code>A</code> wants to paste something into the clipboard, it has to request to the X window server, then become a clipboard owner to send the information back to other applications whenever they send a copy request.</p>
<p>This design was considered natural and often required applications to cooperate: If another application <code>B</code> tries to make a request, to become the next owner of the clipboard, then <code>A</code> will lose its ownership. Afterwards, the copy requests from the application <code>C</code>, <code>D</code>, and so on, will be forwarded to the application <code>B</code> instead of <code>A</code>. Similar to a shared region of memory being overwritten by somebody else and the original owner lost its access.</p>
<p>With the above context information, one can understand that before an application starts to “paste” (serve) the clipboard information, it first obtains the clipboard ownership. Until we get the ownership, the clipboard information will not be available for access purposes.
In other words, if a clipboard API is designed in the following way:</p>
<div><pre><code><span>1</span>clipboard.<span>Write</span>(<span>"some information"</span>)
</code></pre></div>
<p>We have to guarantee from its inside that when the function returns,
the information should be available to be accessed by other applications.</p>
<p>Back then, our first idea to deal with the problem was to pass a channel from Go to C, then send a value through the channel from C to Go. After a quick research, we realized that it is impossible because channels cannot be passed as a value between C and Go due to the <a href="https://pkg.go.dev/cmd/cgo#hdr-Passing_pointers">rules of passing pointers in Cgo</a> (see a previous proposal document <sup></sup><sup></sup>). Even there is a way to pass the entire channel value to the C, there will be no facility to send values through that channel on the C side because C does not have the language support of the <code>&lt;-</code> operator.</p>
<p>The next idea was to pass a function callback, then get it called on the C side. The function's execution will use the desired channel to send a notification back to the waiting goroutine.</p>
<p>After a few attempt, we found that the only possible way is to attach a global function pointer and gets it called through a function wrapper:</p>
<div><pre><code><span> 1</span><span>/*
</span><span> 2</span><span>int myfunc(void* go_value);
</span><span> 3</span><span>*/</span>
<span> 4</span><span>import</span> <span>"C"</span>
<span> 5</span>
<span> 6</span><span>// This funcCallback tries to avoid a runtime panic error when
</span><span> 7</span><span></span><span>// directly pass it to Cgo because it violates the pointer passing
</span><span> 8</span><span></span><span>// rules:
</span><span> 9</span><span></span><span>//
</span><span>10</span><span></span><span>// panic: runtime error: cgo argument has Go pointer to Go pointer
</span><span>11</span><span></span><span>var</span> (
<span>12</span>	funcCallback   <span>func</span>()
<span>13</span>	funcCallbackMu sync.Mutex
<span>14</span>)
<span>15</span>
<span>16</span><span>type</span> gocallback <span>struct</span>{ f <span>func</span>() }
<span>17</span>
<span>18</span><span>func</span> <span>main</span>() {
<span>19</span>	<span>go</span> <span>func</span>() {
<span>20</span>		ret := C.<span>myfunc</span>(unsafe.<span>Pointer</span>(&amp;gocallback{<span>func</span>() {
<span>21</span>			funcCallbackMu.<span>Lock</span>()
<span>22</span>			f := funcCallback <span>// must use a global function variable.
</span><span>23</span><span></span>			funcCallbackMu.<span>Unlock</span>()
<span>24</span>			<span>f</span>()
<span>25</span>		}}))
<span>26</span>		<span>// ... do work ...
</span><span>27</span><span></span>	}()
<span>28</span>	<span>// ... do work ...
</span><span>29</span><span></span>}
</code></pre></div>
<p>In above, the <code>gocallback</code> pointer on the Go side is passed through the C function <code>myfunc</code>. On the C side, there will be a call using <code>go_func_callback</code> that being called on the C, via passing the struct <code>gocallback</code> as a parameter:</p>
<div><pre><code><span>1</span><span>// myfunc will trigger a callback, c_func, whenever it is needed
</span><span>2</span><span></span><span>// and pass the gocallback data though the void* parameter.
</span><span>3</span><span></span><span>void</span> <span>c_func</span>(<span>void</span> *data) {
<span>4</span>	<span>void</span> *gocallback = userData;
<span>5</span>	<span>// the gocallback is received as a pointer, we pass it as
</span><span>6</span><span></span>	<span>// an argument to the go_func_callback
</span><span>7</span><span></span>	go_func_callback(gocallback);
<span>8</span>}
</code></pre></div>
<p>The <code>go_func_callback</code> knows its parameter is typed as <code>gocallback</code>. Thus a type casting is safe to do the call:</p>
<div><pre><code><span>1</span><span>//go:export go_func_callback
</span><span>2</span><span></span><span>func</span> <span>go_func_callback</span>(c unsafe.Pointer) {
<span>3</span>	(*gocallback)(c).<span>call</span>()
<span>4</span>}
<span>5</span>
<span>6</span><span>func</span> (c *gocallback) <span>call</span>() { c.<span>f</span>() }
</code></pre></div>
<p>The function <code>f</code> in the <code>gocallback</code> is exactly what we would like to call:</p>
<div><pre><code><span>1</span><span>func</span>() {
<span>2</span>	funcCallbackMu.<span>Lock</span>()
<span>3</span>	f := funcCallback <span>// must use a global function variable.
</span><span>4</span><span></span>	funcCallbackMu.<span>Unlock</span>()
<span>5</span>	<span>f</span>()               <span>// get called
</span><span>6</span><span></span>}
</code></pre></div>
<p>Note that the <code>funcCallback</code> must be a global function variable. Otherwise, it is a violation of the <a href="https://pkg.go.dev/cmd/cgo/#hdr-Passing_pointers">cgo pointer passing rules</a> as mentioned before.</p>
<p>Furthermore, an immediate reaction to the readability of the above code is: too complicated. The demonstrated approach can only assign one function at a time, which is also a violation of the concurrent nature. Any per-goroutine dedicated application will not benefit from this approach because they need a per-goroutine function callback instead of a single global callback. By then, we wonder if there is a better and elegant approach to deal with it.</p>
<p>Through our research, we found that the need occurs quite often in the community, and is also being proposed in golang/go#37033<sup></sup>. Luckily, such a facility is now ready in Go 1.17 :)</p>
<h2>What is <code>runtime/cgo.Handle</code>?</h2>
<p>The new <a href="https://tip.golang.org/pkg/runtime/cgo/#Handle">runtime/cgo.Handle</a> provides a way to pass values that contain Go pointers (pointers to memory allocated by Go) between Go and C without breaking the cgo pointer passing rules. A <code>Handle</code> is an integer value that can represent any Go value. A <code>Handle</code> can be passed through C and back to Go, and the Go code can use the <code>Handle</code> to retrieve the original Go value. The final API design is proposed as following:</p>
<div><pre><code><span> 1</span><span>package</span> cgo
<span> 2</span>
<span> 3</span><span>type</span> Handle <span>uintptr</span>
<span> 4</span>
<span> 5</span><span>// NewHandle returns a handle for a given value.
</span><span> 6</span><span></span><span>//
</span><span> 7</span><span></span><span>// The handle is valid until the program calls Delete on it.
</span><span> 8</span><span></span><span>// The handle uses resources, and this package assumes that C
</span><span> 9</span><span></span><span>// code may hold on to the handle, so a program must explicitly
</span><span>10</span><span></span><span>// call Delete when the handle is no longer needed.
</span><span>11</span><span></span><span>//
</span><span>12</span><span></span><span>// The intended use is to pass the returned handle to C code,
</span><span>13</span><span></span><span>// which passes it back to Go, which calls Value.
</span><span>14</span><span></span><span>func</span> <span>NewHandle</span>(v <span>interface</span>{}) Handle
<span>15</span>
<span>16</span><span>// Value returns the associated Go value for a valid handle.
</span><span>17</span><span></span><span>//
</span><span>18</span><span></span><span>// The method panics if the handle is invalid.
</span><span>19</span><span></span><span>func</span> (h Handle) <span>Value</span>() <span>interface</span>{}
<span>20</span>
<span>21</span><span>// Delete invalidates a handle. This method should only be
</span><span>22</span><span></span><span>// called once the program no longer needs to pass the handle
</span><span>23</span><span></span><span>// to C and the C code no longer has a copy of the handle value.
</span><span>24</span><span></span><span>//
</span><span>25</span><span></span><span>// The method panics if the handle is invalid.
</span><span>26</span><span></span><span>func</span> (h Handle) <span>Delete</span>()
</code></pre></div>
<p>As we can observe: <code>cgo.NewHandle</code> returns a handle for any given value; the method <code>cgo.(Handle).Value</code> returns the corresponding Go value of the handle; whenever we need to delete the value, one can call <code>cgo.(Handle).Delete</code>.</p>
<p>The most straightforward example is to pass a string between Go and C using <code>Handle</code>. On the Go side:</p>
<div><pre><code><span> 1</span><span>package</span> main
<span> 2</span><span>/*
</span><span> 3</span><span>#include &lt;stdint.h&gt; // for uintptr_t
</span><span> 4</span><span>extern void MyGoPrint(uintptr_t handle);
</span><span> 5</span><span>void myprint(uintptr_t handle);
</span><span> 6</span><span>*/</span>
<span> 7</span><span>import</span> <span>"C"</span>
<span> 8</span><span>import</span> <span>"runtime/cgo"</span>
<span> 9</span>
<span>10</span><span>func</span> <span>main</span>() {
<span>11</span>	s := <span>"Hello The golang.design Initiative"</span>
<span>12</span>	C.<span>myprint</span>(C.<span>uintptr_t</span>(cgo.<span>NewHandle</span>(s)))
<span>13</span>	<span>// Output:
</span><span>14</span><span></span>	<span>// Hello The golang.design Initiative
</span><span>15</span><span></span>}
</code></pre></div>
<p>The string <code>s</code> is passed through a created handle to the C function <code>myprint</code>, and on the C side:</p>
<div><pre><code><span>1</span><span>#</span><span>include</span> <span>&lt;stdint.h&gt; // for uintptr_t</span><span>
</span><span>2</span><span></span>
<span>3</span><span>// A Go function
</span><span>4</span><span></span><span>extern</span> <span>void</span> <span>MyGoPrint</span>(uintptr_t handle);
<span>5</span><span>// A C function
</span><span>6</span><span></span><span>void</span> <span>myprint</span>(uintptr_t handle) {
<span>7</span>	MyGoPrint(handle);
<span>8</span>}
</code></pre></div>
<p>The <code>myprint</code> passes the handle back to a Go function <code>MyGoPrint</code>:</p>
<div><pre><code><span>1</span><span>//go:export MyGoPrint
</span><span>2</span><span></span><span>func</span> <span>MyGoPrint</span>(handle C.uintptr_t) {
<span>3</span>	h := cgo.<span>Handle</span>(handle)
<span>4</span>	s := h.<span>Value</span>().(<span>string</span>)
<span>5</span>	<span>println</span>(s)
<span>6</span>	h.<span>Delete</span>()
<span>7</span>}
</code></pre></div>
<p>The <code>MyGoPrint</code> queries the value using <code>cgo.(Handle).Value()</code> and prints it out. Then deletes the value using <code>cgo.(Handle).Delete()</code>.</p>
<p>With this new facility, we can simplify the previously mentioned function
callback pattern much better:</p>
<div><pre><code><span> 1</span><span>/*
</span><span> 2</span><span>#include &lt;stdint.h&gt;
</span><span> 3</span><span>
</span><span> 4</span><span>int myfunc(void* go_value);
</span><span> 5</span><span>*/</span>
<span> 6</span><span>import</span> <span>"C"</span>
<span> 7</span>
<span> 8</span><span>func</span> <span>main</span>() {
<span> 9</span>
<span>10</span>	ch := <span>make</span>(<span>chan</span> <span>struct</span>{})
<span>11</span>	handle := cgo.<span>NewHandle</span>(ch)
<span>12</span>	<span>go</span> <span>func</span>() {
<span>13</span>		<span>// myfunc will call goCallback when needed.
</span><span>14</span><span></span>		C.<span>myfunc</span>(C.<span>uintptr_t</span>(handle))
<span>15</span>		...
<span>16</span>	}()
<span>17</span>
<span>18</span>	&lt;-ch <span>// we got notified from the myfunc.
</span><span>19</span><span></span>	handle.<span>Delete</span>() <span>// no need thus delete the handle.
</span><span>20</span><span></span>	...
<span>21</span>}
<span>22</span>
<span>23</span><span>//go:export goCallback
</span><span>24</span><span></span><span>func</span> <span>goCallback</span>(h C.uintptr_t) {
<span>25</span>	v := cgo.<span>Handle</span>(h).<span>Value</span>().(<span>chan</span> <span>struct</span>{})
<span>26</span>	v &lt;- <span>struct</span>{}
<span>27</span>}
</code></pre></div>
<p>More importantly, the <code>cgo.Handle</code> is a concurrent-safe mechanism, which means that once we have the handle number, we can fetch the value (if still available) anywhere without suffering from data race.</p>
<p>Next question: How to implement <code>cgo.Handle</code>?</p>
<h2>First Attempt</h2>
<p>The first attempt<sup></sup> was a lot complicated. Since we need a centralized way to manage all pointers in a concurrent-safe way, the quickest idea that comes to our mind was the <code>sync.Map</code> that maps a unique number to the desired value. Hence, we can easily use a global <code>sync.Map</code>:</p>
<div><pre><code><span>1</span><span>package</span> cgo
<span>2</span>
<span>3</span><span>var</span> m = &amp;sync.Map{}
</code></pre></div>
<p>However, we have to think about the core challenge:
How to allocate a runtime-level unique ID? Passing an integer between Go and C is relatively easy, what could be a unique representation for a given value?</p>
<p>The first idea is the memory address. Because every pointer or value
is stored somewhere in memory, if we can have the information, it would
be very easy to use as the ID of the value because each value has exactly one unique memory address.</p>
<p>To complete this idea, we need to be a little bit cautious: Will the memory address of a living value is changed at some point? The question leads to two more questions:</p>
<ol>
<li>What if a value is on the goroutine stack? If so, the value will be released when the goroutine is dead.</li>
<li>Go is a garbage-collected language. What if the garbage collector moves and compacts the value to a different place? Then the memory address of the value will be changed, too.</li>
</ol>
<p>Based on our years of <a href="https://golang.design/s/more">experience and understanding</a> of the runtime, we learned that the Go's garbage collector before 1.17 is always not moving and the mechanism is also very unlikely to change. That means, if a value is living on the heap, it will not be moved to other places. With this fact, we are good with the second question.</p>
<p>It is a little bit tricky for the first question: a value on the stack may move as the stack grows. The more intractable part is that compiler optimization may move values between stacks, and runtime may move the stack when the stack ran out of its size.</p>
<p>Naturally, we might ask: is it possible to make sure a value always be allocated on the heap instead of the stack? The answer is: Yes! If we turn it into an <code>interface{}</code>. Until 1.17, the Go compiler's escape analysis always marks the value that should escape to the heap if it is converted as an <code>interface{}</code>.</p>
<p>With all the analysis above, we can write the following part of the implementation that utilizes the memory address of an escaped value:</p>
<div><pre><code><span> 1</span><span>// wrap wraps a Go value.
</span><span> 2</span><span></span><span>type</span> wrap <span>struct</span>{ v <span>interface</span>{} }
<span> 3</span>
<span> 4</span><span>func</span> <span>NewHandle</span>(v <span>interface</span>{}) Handle {
<span> 5</span>	<span>var</span> k <span>uintptr</span>
<span> 6</span>
<span> 7</span>	rv := reflect.<span>ValueOf</span>(v)
<span> 8</span>	<span>switch</span> rv.<span>Kind</span>() {
<span> 9</span>	<span>case</span> reflect.Ptr, reflect.UnsafePointer, reflect.Slice,
<span>10</span>		reflect.Map, reflect.Chan, reflect.Func:
<span>11</span>		<span>if</span> rv.<span>IsNil</span>() {
<span>12</span>			<span>panic</span>(<span>"cgo: cannot use Handle for nil value"</span>)
<span>13</span>		}
<span>14</span>
<span>15</span>		k = rv.<span>Pointer</span>()
<span>16</span>	<span>default</span>:
<span>17</span>		<span>// Wrap and turn a value parameter into a pointer.
</span><span>18</span><span></span>		<span>// This enables us to always store the passing object
</span><span>19</span><span></span>		<span>// as a pointer, and helps to identify which of whose
</span><span>20</span><span></span>		<span>// are initially pointers or values when Value is called.
</span><span>21</span><span></span>		v = &amp;wrap{v}
<span>22</span>		k = reflect.<span>ValueOf</span>(v).<span>Pointer</span>()
<span>23</span>	}
<span>24</span>
<span>25</span>	...
<span>26</span>}
</code></pre></div>
<p>Note that the implementation above treats the values differently: For <code>reflect.Ptr</code>, <code>reflect.UnsafePointer</code>, <code>reflect.Slice</code>, <code>reflect.Map</code>, <code>reflect.Chan</code>, <code>reflect.Func</code> types, they are already pointers escaped to the heap, we can safely get the address from them. For the other kinds, we need to turn them from a value to a pointer and also make sure they will always escape to the heap. That is the part:</p>
<div><pre><code><span>1</span>		<span>// Wrap and turn a value parameter into a pointer. This
</span><span>2</span><span></span>		<span>// enables us to always store the passing object as a
</span><span>3</span><span></span>		<span>// pointer, and helps to identify which of whose are
</span><span>4</span><span></span>		<span>// initially pointers or values when Value is called.
</span><span>5</span><span></span>		v = &amp;wrap{v}
<span>6</span>		k = reflect.<span>ValueOf</span>(v).<span>Pointer</span>()
</code></pre></div>
<p>Now we have turned everything into an escaped value on the heap. The next thing we have to ask is: what if the two values are the same? That means the <code>v</code> passed to <code>cgo.NewHandle(v)</code> is the same object. Then we will get the same memory address in <code>k</code> at this point.</p>
<p>The easy case is, of course, if the address is not on the global map, then we do not have to think but return the address as the handle of the value:</p>
<div><pre><code><span> 1</span><span>func</span> <span>NewHandle</span>(v <span>interface</span>{}) Handle {
<span> 2</span>	...
<span> 3</span>
<span> 4</span>	<span>// v was escaped to the heap because of reflection. As Go do
</span><span> 5</span><span></span>	<span>// not have a moving GC (and possibly lasts true for a long
</span><span> 6</span><span></span>	<span>// future), it is safe to use its pointer address as the key
</span><span> 7</span><span></span>	<span>// of the global map at this moment. The implementation must
</span><span> 8</span><span></span>	<span>// be reconsidered if moving GC is introduced internally in
</span><span> 9</span><span></span>	<span>// the runtime.
</span><span>10</span><span></span>	actual, loaded := m.<span>LoadOrStore</span>(k, v)
<span>11</span>	<span>if</span> !loaded {
<span>12</span>	    <span>return</span> <span>Handle</span>(k)
<span>13</span>	}
<span>14</span>
<span>15</span>	...
<span>16</span>}
</code></pre></div>
<p>Otherwise, we have to check the old value in the global map, if it is the same value, then we return the same address as expected:</p>
<div><pre><code><span> 1</span><span>func</span> <span>NewHandle</span>(v <span>interface</span>{}) Handle {
<span> 2</span>	...
<span> 3</span>
<span> 4</span>	arv := reflect.<span>ValueOf</span>(actual)
<span> 5</span>	<span>switch</span> arv.<span>Kind</span>() {
<span> 6</span>	<span>case</span> reflect.Ptr, reflect.UnsafePointer, reflect.Slice,
<span> 7</span>		reflect.Map, reflect.Chan, reflect.Func:
<span> 8</span>		<span>// The underlying object of the given Go value already have
</span><span> 9</span><span></span>		<span>// its existing handle.
</span><span>10</span><span></span>		<span>if</span> arv.<span>Pointer</span>() == k {
<span>11</span>			<span>return</span> <span>Handle</span>(k)
<span>12</span>		}
<span>13</span>
<span>14</span>		<span>// If the loaded pointer is inconsistent with the new
</span><span>15</span><span></span>		<span>// pointer, it means the address has been used for
</span><span>16</span><span></span>		<span>// different objects because of GC and its address is
</span><span>17</span><span></span>		<span>// reused for a new Go object, meaning that the Handle
</span><span>18</span><span></span>		<span>// does not call Delete explicitly when the old Go value
</span><span>19</span><span></span>		<span>// is not needed. Consider this as a misuse of a handle,
</span><span>20</span><span></span>		<span>// do panic.
</span><span>21</span><span></span>		<span>panic</span>(<span>"cgo: misuse of a Handle"</span>)
<span>22</span>	<span>default</span>:
<span>23</span>		<span>panic</span>(<span>"cgo: Handle implementation has an internal bug"</span>)
<span>24</span>	}
<span>25</span>}
</code></pre></div>
<p>If the existing value shares the same address with the newly requested
value, this must be a misuse of the Handle.</p>
<p>Since we have used the <code>wrap</code> struct to turn everything into the <code>reflect.Ptr</code> type, it is impossible to have other kinds of values to fetch from the global map. If that happens, it is an internal bug in the handle implementation.</p>
<p>When implementing the <code>Value()</code> method, we see why a <code>wrap</code> struct beneficial:</p>
<div><pre><code><span> 1</span><span>func</span> (h Handle) <span>Value</span>() <span>interface</span>{} {
<span> 2</span>	v, ok := m.<span>Load</span>(<span>uintptr</span>(h))
<span> 3</span>	<span>if</span> !ok {
<span> 4</span>		<span>panic</span>(<span>"cgo: misuse of an invalid Handle"</span>)
<span> 5</span>	}
<span> 6</span>	<span>if</span> wv, ok := v.(*wrap); ok {
<span> 7</span>		<span>return</span> wv.v
<span> 8</span>	}
<span> 9</span>	<span>return</span> v
<span>10</span>}
</code></pre></div>
<p>Because we can check when the stored object is a <code>*wrap</code> pointer, which means it was a value other than pointers. We return the value instead of the stored object.</p>
<p>Lastly, the <code>Delete</code> method becomes trivial:</p>
<div><pre><code><span>1</span><span>func</span> (h Handle) <span>Delete</span>() {
<span>2</span>	_, ok := m.<span>LoadAndDelete</span>(<span>uintptr</span>(h))
<span>3</span>	<span>if</span> !ok {
<span>4</span>		<span>panic</span>(<span>"cgo: misuse of an invalid Handle"</span>)
<span>5</span>	}
<span>6</span>}
</code></pre></div>
<p>See a full implementation in <a href="https://github.com/golang-design/clipboard/blob/main/internal/cgo/handle.go">golang.design/x/clipboard/internal/cgo</a>.</p>
<h2>The Accepted Approach</h2>
<p>As one may have realized, the previous approach is much more complicated than expected and non-trivial: it relies on the foundation that runtime garbage collector is not a moving garbage collector, and an argument though interfaces will escape to the heap.</p>
<p>Although several other places in the internal runtime implementation rely on these facts, such as the channel implementation, it is still a little over-complicated than what we expected.</p>
<p>Notably, the previous <code>NewHandle</code> actually behaves to return a unique handle when the provided Go value refers to the same object. This is the core that brings the complexity of the implementation. However, we have another possibility: <code>NewHandle</code> always returns a different handle, and a Go value can have multiple handles.</p>
<p>Do we really need to Handle to be unique and keep it satisfy <a href="https://en.wikipedia.org/wiki/Idempotence">idempotence</a>? After a short discussion with the Go team, we share the consensus that for the purpose of a Handle, it seems unnecessary to keep it unique for the following reasons:</p>
<ol>
<li>The semantic of <code>NewHandle</code> is to return a <em>new</em> handle, instead of a unique handle;</li>
<li>The handle is nothing more than just an integer and guarantee it to be unique may prevent misuse of the handle, but it cannot always avoid the misuse until it is too late;</li>
<li>The complexity of the implementation.</li>
</ol>
<p>Therefore, we need to rethink the original question: How to allocate a runtime-level unique ID?</p>
<p>In reality, the approach is more manageable: we only need to increase a number and never stop. This is the most commonly used approach for unique ID generation. For instance, in database applications, the unique id of a table row is always incremental; in Unix timestamp, the time is always incremental, etc.</p>
<p>If we use the same approach, what would be a possible concurrent-safe implementation? With <code>sync.Map</code> and atomic, we can produce code like this:</p>
<div><pre><code><span> 1</span><span>func</span> <span>NewHandle</span>(v <span>interface</span>{}) Handle {
<span> 2</span>	h := atomic.<span>AddUintptr</span>(&amp;handleIdx, <span>1</span>)
<span> 3</span>	<span>if</span> h == <span>0</span> {
<span> 4</span>		<span>panic</span>(<span>"runtime/cgo: ran out of handle space"</span>)
<span> 5</span>	}
<span> 6</span>
<span> 7</span>	handles.<span>Store</span>(h, v)
<span> 8</span>	<span>return</span> <span>Handle</span>(h)
<span> 9</span>}
<span>10</span>
<span>11</span><span>var</span> (
<span>12</span>	handles   = sync.Map{} <span>// map[Handle]interface{}
</span><span>13</span><span></span>	handleIdx <span>uintptr</span>      <span>// atomic
</span><span>14</span><span></span>)
</code></pre></div>
<p>Whenever we want to allocate a new ID (<code>NewHandle</code>), one can increase the handle number <code>handleIdx</code> atomically, then the next allocation will always be guaranteed to have a larger number to use. With that allocated number, we can easily store it to a global map that persists all the Go values.</p>
<p>The remaining work becomes trivial. When we want to use the handle to retrieve the corresponding Go value back, we access the value map via the handle number:</p>
<div><pre><code><span>1</span><span>func</span> (h Handle) <span>Value</span>() <span>interface</span>{} {
<span>2</span>	v, ok := handles.<span>Load</span>(<span>uintptr</span>(h))
<span>3</span>	<span>if</span> !ok {
<span>4</span>		<span>panic</span>(<span>"runtime/cgo: misuse of an invalid Handle"</span>)
<span>5</span>	}
<span>6</span>	<span>return</span> v
<span>7</span>}
</code></pre></div>
<p>Further, if we are done with the handle, one can delete it from the value map:</p>
<div><pre><code><span>1</span><span>func</span> (h Handle) <span>Delete</span>() {
<span>2</span>	_, ok := handles.<span>LoadAndDelete</span>(<span>uintptr</span>(h))
<span>3</span>	<span>if</span> !ok {
<span>4</span>		<span>panic</span>(<span>"runtime/cgo: misuse of an invalid Handle"</span>)
<span>5</span>	}
<span>6</span>}
</code></pre></div>
<p>In this implementation, we do not have to assume the runtime mechanism but just use the language. As long as the Go 1 compatibility keeps the promise <code>sync.Map</code> to work, there will be no need to rework the whole <code>Handle</code> design. Because of its simplicity, this is the accepted approach (see CL 295369<sup></sup>) by the Go team.</p>
<p>Aside from a future re-implementation of <code>sync.Map</code> that optimizes parallelism, the <code>Handle</code> will automatically benefit from it. Let us do a final benchmark that compares the previous method and the current approach:</p>
<div><pre><code><span> 1</span><span>func</span> <span>BenchmarkHandle</span>(b *testing.B) {
<span> 2</span>	b.<span>Run</span>(<span>"non-concurrent"</span>, <span>func</span>(b *testing.B) {
<span> 3</span>		<span>for</span> i := <span>0</span>; i &lt; b.N; i++ {
<span> 4</span>			h := cgo.<span>NewHandle</span>(i)
<span> 5</span>			_ = h.<span>Value</span>()
<span> 6</span>			h.<span>Delete</span>()
<span> 7</span>	    }
<span> 8</span>	})
<span> 9</span>	b.<span>Run</span>(<span>"concurrent"</span>, <span>func</span>(b *testing.B) {
<span>10</span>		b.<span>RunParallel</span>(<span>func</span>(pb *testing.PB) {
<span>11</span>			<span>var</span> v <span>int</span>
<span>12</span>			<span>for</span> pb.<span>Next</span>() {
<span>13</span>				h := cgo.<span>NewHandle</span>(v)
<span>14</span>				_ = h.<span>Value</span>()
<span>15</span>				h.<span>Delete</span>()
<span>16</span>			}
<span>17</span>		})
<span>18</span>	})
<span>19</span>}
</code></pre></div>
<pre><code>name                     old time/op  new time/op  delta
Handle/non-concurrent-8  407ns ±1%    393ns ±2%   -3.51%  (p=0.000 n=8+9)
Handle/concurrent-8      768ns ±0%    759ns ±1%   -1.21%  (p=0.003 n=9+9)
</code></pre>
<p>Simpler, faster, why not?</p>
<h2>Conclusion</h2>
<p>This article discussed the newly introduced <code>runtime/cgo.Handle</code> facility coming in the Go 1.17 release that we contributed. The <code>Handle</code> facility enables us to pass Go values between Go and C back and forth without breaking the cgo pointer passing rules. After a short introduction to the usage of the feature, we first discussed a first attempt implementation based on the fact that the runtime garbage collector is not a moving GC and the escape behavior of <code>interface{}</code> arguments.
After a few discussions of the ambiguity of the Handle semantics and the drawbacks in the previous implementation, we also introduced a straightforward and better-performed approach and demonstrated its performance.</p>
<p>As a real-world demonstration, we have been using the mentioned two approaches
in two of our released packages for quite a long time:
<a href="https://github.com/golang-design/clipboard">golang.design/x/clipboard</a>
and <a href="https://github.com/golang-design/hotkey">golang.design/x/hotkey</a> <sup></sup>
before in their <code>internal/cgo</code> package.
We are looking forward to switching to the officially released <code>runtime/cgo</code>
package in the Go 1.17 release.</p>
<p>For future work, one can foresee that a possible limitation in the accepted
implementation is that the handle number may run out of the handle space
very quickly in 32-bit or lower operating systems (similar to
<a href="https://en.wikipedia.org/wiki/Year_2038_problem">Year 2038 Problem</a>.
When we allocate 100 handles per second, the handle space can run out in
0xFFFFFFF / (24 * 60 * 60 * 100) = 31 days).</p>
<p>*<em>If you are interested and think this is a serious issue, feel free to
<a href="https://golang.design/cdn-cgi/l/email-protection#3850511d0d7a594c1d0d7c5f575459565f165c5d4b515f56">CC us</a> when you send a CL,
it would also be interesting for us to read your excellent approach.</em></p>
<h2>References</h2>
</div>
</div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via The golang.design Initiative <br>
<br>
April 28, 2023 at 05:34PM
